include/swift/AST/DiagnosticsSema.def

```diff
@@ -2446,6 +2446,9 @@ ERROR(broken_vector_numeric_requirement,none,
       "VectorNumeric protocol is broken: unexpected requirement", ())
 ERROR(broken_differentiable_requirement,none,
       "Differentiable protocol is broken: unexpected requirement", ())
+ERROR(differentiable_need_explicit_addarith_conformance,none,
+      "synthesizing a conformance to 'Differentiable' requires an explicit "
+      "%0 : 'AdditiveArithmetic' conformance constraint", (Type))
```

Comment:
It is a bit unnatural to leave `:` out of quotation marks. It would be better to put the entire conformance requirement in a single pair of quotation marks.

```console
synthesizing a conformance to 'Differentiable' requires an explicit 'T.TangentVector : AdditiveArithmetic' conformance constraint
```

Can you programmatically synthesize a `Requirement` and just print that?

Reply:
Good point! Unfortunately, I don't think `DiagnosticEngine` has support for printing `Requirement`s yet:
```
// include/swift/AST/DiagnosticsSema.def
ERROR(differentiable_need_explicit_addarith_conformance,none,
      "synthesizing a conformance to 'Differentiable' requires an explicit "
      "%0 conformance constraint", (Requirement))

// Error:
/Users/danielzheng/swift-dev/swift/include/swift/AST/DiagnosticEngine.h:316:32: error: no viable conversion from 'typename remove_reference<Requirement &>::type' (aka 'swift::Requirement') to 'swift::DiagnosticArgument'
        DiagnosticArgument(0), std::move(VArgs)...
                               ^~~~~~~~~~~~~~~~
```

I actually copied existing diagnostic messages regarding conformances, which don't print entire conformances within single quotes. Example:
```
sr9595.swift:8:22: warning: redundant conformance constraint 'Self.T': 'P'
  associatedtype T : P & Q where T.Assoc : CaseIterable, T == U.U
                     ^
```

It would be nice to improve this in the future though. I'd prefer not to look into it now, if that's okay.

Reply:
Why not just use a `StringRef` and print it to a string?

Reply:
Very nice!